### PR TITLE
Trace cli in go

### PR DESCRIPTION
**Pull request by turkenh**
PR for https://github.com/crossplaneio/crossplane/issues/744

To do:

- [x] Test and adapt "Simple Resource Class Selection" https://github.com/crossplaneio/crossplane/pull/926
- [x] Better identifier, UID missing if object does not exist
- [x] Adapt bootstrap.sh to include trace
- [x] Ready/Pending/Missing states to highlight what is wrong
- [x] Unit tests
- [x] Test different scenarios and edge case handling 
- [ ] Refactor: better naming, using constants, comments etc
- [x] Polish graph output

Build & Install:
```
$ make build
$ make install
```

Trace a kubernetesapplication:
```
$ kubectl crossplane trace kubernetesapplication <name> -n <namespace>
```

Graph output:
```
# Graphviz needs to be installed first, e.g. brew install graphviz
$  kubectl crossplane trace kubernetesapplication <name> -n <namespace> -o dot |dot -Tpng > /tmp/output.png&& open /tmp/output.png
```

### CODE REVIEW

File: bin/kubectl-crossplane-trace [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+set -e
+
+function usage {
+  echo "Usage: kubectl crossplane trace KIND NAME [-n| --namespace NAMESPACE] [-h|--help]" >&2 >&2
```
Comment (suskin) [inline bin/kubectl-crossplane-trace]: It seems like `>&2 >&2` should probably be `>&2`

File: bin/kubectl-crossplane-trace [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+function check_help {
+  if [ "${1}" == "-h" -o "${1}" == "--help" ] ; then
+    usage
+    exit 1
+  fi
+}
+
+check_help "${1}"
+
+crossplane-cli trace "${@}"
```
Comment (suskin) [inline bin/kubectl-crossplane-trace]: Where does `crossplane-cli` come from?

File: bin/kubectl-crossplane-trace [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
```
Comment (suskin) [inline bin/kubectl-crossplane-trace]: What is the benefit of having a `bash` wrapper around another binary?

**Review by suskin — COMMENTED** [review]
@turkenh I like that there's an early draft for early review and comments! A couple questions.

Would you be able to use more upbound-style commit messages? Here's the formula I've been using:

```
category: one-line summary of change

This section down here is longer, and is an explanation
of why the change is being made, or why it is happening
in the way it is being done. The goal of this section is to
help people get into the mindset of the original author, so
that they can keep making changes in the way that the
original author would be, if that's what they want to do.

commit signature
```

What testing has been done?

File: bin/kubectl-crossplane-trace (hunk 2 above)
Comment (turkenh) [reply bin/kubectl-crossplane-trace]: This will be the built binary of go code. 
e.g.
```
go build -o /bin/crossplane-cli cmd/main.go
```

File: bin/kubectl-crossplane-trace (hunk 3 above)
Comment (turkenh) [reply bin/kubectl-crossplane-trace]: As you know, kubectl plugins require an exact name matching the command/subcommand.
This way, we can use a single binary produced from the go code for different subcommands (currently only for trace but later we can consider adding more).

**Comment by turkenh** [comment]
> Would you be able to use more upbound-style commit messages? Here's the formula I've been using

Sure. I'll most probably, overwrite the commit history here with some sort of "Initial commit" for trace cli go project since the initial commits include experimental changes and I didn't follow well scoped commits.

> What testing has been done?

I am mainly testing with wordpress stack example both happy path and also by intentionally breaking some stuff.

**Comment by turkenh** [comment]
@suskin I still have something to do, but believe that PR is ready for review.

It would be great if you could have a chance to test on your side as well.

File: bootstrap.sh [hunk 4]
```diff
@@ -10,13 +10,23 @@ if [[ -z "${RELEASE}" ]]; then
   RELEASE=master
 fi
 
+PLATFORM="linux"
```
Comment (suskin) [inline bootstrap.sh:13]: @turkenh what platform are you running? I'm testing this on darwin.

File: Makefile [hunk 5]
```diff
@@ -4,8 +4,16 @@ default:
...
+clean:
+	rm bin/kubectl-crossplane-trace
+.PHONY: clean
+
 install:
-	ln -si $(abspath bin/kubectl-crossplane-stack-*) $(INSTALL_DIR)/
+	ln -si $(abspath bin/kubectl-crossplane-*) $(INSTALL_DIR)/
 .PHONY: install
 
 uninstall:
```
Comment (suskin) [inline Makefile:23]: Is the `.DS_Store` file below needed? Should we remove it and .gitignore it?

File: bootstrap.sh [hunk 6]
```diff
@@ -10,13 +10,23 @@ if [[ -z "${RELEASE}" ]]; then
...
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-publish https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-publish >/dev/null
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-install https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-install >/dev/null
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-uninstall https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-uninstall >/dev/null
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-generate_install https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-generate_install >/dev/null
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-list https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-list >/dev/null
+else
+  curl -sL https://github.com/crossplaneio/crossplane-cli/releases/download/"${RELEASE}"/crossplane-cli-"${RELEASE}"-"${PLATFORM}".tar.gz | tar xz --strip 1 -C "${PREFIX}"/bin
```
Comment (suskin) [inline bootstrap.sh:30]: Has this been tested on linux and darwin?

File: bootstrap.sh [hunk 7]
```diff
@@ -10,13 +10,23 @@ if [[ -z "${RELEASE}" ]]; then
...
-curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-publish https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-publish >/dev/null
-curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-install https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-install >/dev/null
-curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-uninstall https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-uninstall >/dev/null
-curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-generate_install https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-generate_install >/dev/null
-curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-list https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-list >/dev/null
+if [[ "${RELEASE}" == "master" ]]; then
+  echo "trace subcommand will not be available from master, RELEASE must be set to a released version."
```
Comment (suskin) [inline bootstrap.sh]: I have been taking the approach of sending progress and error messages to stderr instead; let's redirect this: `>&2`

Also, could we put some sort of attention-grabbing prefix at the beginning of the message, and add an example of a released version? For example:

```
echo "NOTICE: the trace command is not available from master. RELEASE must be set to a released version (such as v0.2.0). See https://github.com/crossplaneio/crossplane-cli/releases for the full list of releases." >&2
```

File: bootstrap.sh [hunk 8]
```diff
@@ -10,13 +10,23 @@ if [[ -z "${RELEASE}" ]]; then
   RELEASE=master
 fi
 
+PLATFORM="linux"
+if [[ "$OSTYPE" == "darwin"* ]]; then
```
Comment (suskin) [inline bootstrap.sh]: Style nitpick: could we change this to use curly braces for consistency? So something more like `"${OSTYPE}"`. I like using curly braces for clarity for scripts. In general, I like to use the long forms of syntax in scripts, since they're usually clearer.

File: bootstrap.sh [hunk 9]
```diff
@@ -10,13 +10,23 @@ if [[ -z "${RELEASE}" ]]; then
...
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-publish https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-publish >/dev/null
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-install https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-install >/dev/null
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-uninstall https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-uninstall >/dev/null
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-generate_install https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-generate_install >/dev/null
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-list https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-list >/dev/null
+else
+  curl -sL https://github.com/crossplaneio/crossplane-cli/releases/download/"${RELEASE}"/crossplane-cli-"${RELEASE}"-"${PLATFORM}".tar.gz | tar xz --strip 1 -C "${PREFIX}"/bin
+fi
```
Comment (suskin) [inline bootstrap.sh:31]: Style nitpick: I would normally put a blank line after an `if` block

File: cmd/trace/main.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+)
+
+func main() {
+	var help bool
+	var kubeconfig string
+	var namespace string
+	var outputFormat string
+	if home := homedir.HomeDir(); home != "" {
+		pflag.StringVar(&kubeconfig, "kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) Absolute path to the kubeconfig file")
```
Comment (suskin) [inline cmd/trace/main.go]: I don't know enough about CLI libraries in go enough to have an opinion here, but is this the library that kubectl uses? I would default to trying to use the same stuff as kubectl.

File: cmd/trace/main.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+	}
+}
+
+func failWithMessage(msg string) {
+	fmt.Fprintln(os.Stderr, msg)
+	os.Exit(-1)
+}
+
+func printHelp() {
+	fmt.Fprintf(os.Stderr, `Usage: kubectl crossplane trace TYPE[.GROUP] NAME [-n| --namespace NAMESPACE] [-h|--help]
```
Comment (suskin) [inline cmd/trace/main.go:108]: Usability: it'd probably make sense to also include `-o`  in the usage line, or something like `[options ...]`

File: cmd/trace/main.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+	if help {
+		printHelp()
+		os.Exit(0)
+	}
+
+	resource := pflag.Arg(0)
+	resourceName := pflag.Arg(1)
+	if resource == "" || resourceName == "" {
+		failWithMessage("missing arguments: SUBCOMMAND TYPE[.GROUP] NAME [-n|--namespace NAMESPACE]")
```
Comment (suskin) [inline cmd/trace/main.go:54]: I'm noticing this message doesn't match the one printed by usage . . . would it possible for us to reuse the same message? What do you think of the pattern of printing the full usage when the arguments are wrong? That's usually the pattern I go with

File: cmd/trace/main.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	fmt.Fprintf(os.Stderr, "Collecting information for %s %s in namespace %s...\n\n", resource, resourceName, namespace)
```
Comment (suskin) [inline cmd/trace/main.go:56]: Nice use of stderr here

File: pkg/crossplane/application.go [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+	fieldsAppDesiredRes          = append(fieldsStatus, "desiredResources")
+	fieldsAppSubmittedRes        = append(fieldsStatus, "submittedResources")
+	fieldsAppResourceSelector    = append(fieldsSpec, "resourceSelector")
+	fieldsAppResourceMatchLabels = append(fieldsAppResourceSelector, "matchLabels")
+
+	stateAppSubmitted = "Submitted"
+)
+
+type Application struct {
+	u *unstructured.Unstructured
```
Comment (suskin) [inline pkg/crossplane/application.go]: In general I like descriptive variable names over single-letter names, but if this is the go idiom for untyped and anonymous variables like this, I'm okay with it.

File: pkg/crossplane/application.go [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+		return ObjectDetails{}
+	}
+	return getObjectDetails(o.u)
+}
+
+func (o *Application) IsReady() bool {
+	return o.GetStatus() == stateAppSubmitted
+}
+
+func (o *Application) GetRelated(filterByLabel func(metav1.GroupVersionKind, string, string) ([]unstructured.Unstructured, error)) ([]*unstructured.Unstructured, error) {
```
Comment (suskin) [inline pkg/crossplane/application.go:47]: Something to keep in mind for the future is that it'd be great if we could add support for more types without having to recompile the `trace` binary. We don't have to have a solution now, but it's something to think about.

File: pkg/crossplane/application_test.go [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+	"testing"
+
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	"github.com/google/go-cmp/cmp"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func TestApplication_GetObjectDetails(t *testing.T) {
```
Comment (suskin) [inline pkg/crossplane/application_test.go:13]: Awesome! Could we also add a `make test` target for running the unit tests?

File: pkg/crossplane/application.go [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+type Application struct {
+	u *unstructured.Unstructured
+}
+
+func NewApplication(u *unstructured.Unstructured) *Application {
+	return &Application{u: u}
+}
+
+func (o *Application) GetStatus() string {
+	return getNestedString(o.u.Object, fieldsStatusState...)
```
Comment (suskin) [inline pkg/crossplane/application.go]: I usually prefer that the imports for shared functionality are explicit so they're easier to follow for readers who are not familiar with the codebase, but this is something we can probably reexamine in the future.

File: pkg/trace/graphprinter.go [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+	writer io.Writer
+}
+
+// NewGraphPrinter returns a GraphPrinter
+func NewGraphPrinter() *GraphPrinter {
+	return &GraphPrinter{writer: os.Stdout}
+}
+
+// Print prints graph definition in Graphviz dot format for a given root node.
+func (p *GraphPrinter) Print(nodes []*Node) error {
```
Comment (suskin) [inline pkg/trace/graphprinter.go:22]: Yesss

**Review by suskin — APPROVED** [review]
Awesome - thanks @turkenh ! Since it's a larger pull request it's a little harder to do an in-depth review quickly, but it looks pretty good to me. I just had some minor questions and suggestions.

I'm going to approve this since we're in different timezones, but I'd like us to at a minimum address the comments about testing (add `make test`, verify that installation from a release has been tested) and usability before we merge the changes.

For posterity, I asked @turkenh in a separate conversation whether we would be able to use more of the code from kubectl - for example, for the `kubectl get` commands. He said that he looked, but that he wasn't able to find anything which was easily reusable. We both agreed that we'd like to revisit that in the future at some point, but that we're okay with doing a quick thing for now so we can start using the tool.

**Comment by suskin** [comment]
I forgot to mention, but I did some integration testing on my side as well. I used it to inspect the statuses for a mysqlinstance and a kubernetesapplication which I created as part of the Stacks guide. And I used it to watch the kubernetesapplication become ready over time. It was way better than doing those things without the tool!

As another note, we'll want to update some of the documentation - in [this repo](https://github.com/crossplaneio/crossplane-cli/blob/master/README.md) and in [crossplane](https://github.com/crossplaneio/crossplane/blame/518648b051cd2055fb2d95dbdbc120855af7eedf/docs/stacks-guide.md#L202) - to recommend that we use `trace`. The documentation will need to be updated before we consider the `trace` work to be complete.

File: bootstrap.sh (hunk 4 above)
Comment (turkenh) [reply bootstrap.sh:13]: I'm also using darwin, but just tested on a debian instance and working fine.

File: Makefile (hunk 5 above)
Comment (turkenh) [reply Makefile:23]: No, definitely not needed. Deleted and added to .gitignore.

File: bootstrap.sh (hunk 6 above)
Comment (turkenh) [reply bootstrap.sh:30]: Yes, works fine on both. Tested with my test releases: https://github.com/turkenh/crossplane-cli/releases/tag/v0.1.2-test

We need to follow the naming convention here with our releases.

`crossplane-cli-"${RELEASE}"-"${PLATFORM}".tar.gz`

File: bootstrap.sh (hunk 7 above)
Comment (turkenh) [reply bootstrap.sh]: Both makes sense. Done.

File: pkg/crossplane/application.go (hunk 14 above)
Comment (turkenh) [reply pkg/crossplane/application.go]: Go idiom likes short names but not sure this applies here. 
Changed it as u -> instance which is already something that I used somewhere else in the code.

File: pkg/crossplane/application.go (hunk 15 above)
Comment (turkenh) [reply pkg/crossplane/application.go:47]: Yes, that's a good point.
This is something that I also considered in the context of making this tool as a "general purpose" tracing tool as you mentioned in the https://github.com/crossplaneio/crossplane/issues/744. This could be possible with some sort of configuration file which describes the types and their relations with other types.

File: pkg/crossplane/application.go (hunk 17 above)
Comment (turkenh) [reply pkg/crossplane/application.go]: Well, if you mean `getNestedString` this is not coming from an import. it is a helper in the same package.
Otherwise, I agree with your comment in general.

**Comment by turkenh** [comment]
@suskin Thanks a lot for reviewing and testing!
All of your comments should be resolved right now.

> For posterity, I asked @turkenh in a separate conversation whether we would be able to use more of the code from kubectl - for example, for the kubectl get commands. He said that he looked, but that he wasn't able to find anything which was easily reusable. We both agreed that we'd like to revisit that in the future at some point, but that we're okay with doing a quick thing for now so we can start using the tool.

Most related parts from `kubectl get` was the ones for printing, since for interacting with api-server I used client-go dynamic client which was pretty good. I checked https://github.com/kubernetes/cli-runtime which could be helpful, but as @suskin mentioned, I found adding my own simple printer as more easy and flexible. I still agree that this should be something to revisit in future.

> I forgot to mention, but I did some integration testing on my side as well. I used it to inspect the statuses for a mysqlinstance and a kubernetesapplication which I created as part of the Stacks guide. And I used it to watch the kubernetesapplication become ready over time. It was way better than doing those things without the tool!

Thanks a lot!
Which cloud provider were you using? I always used GCP and now looking for a way to test with others. Another this that needs to be test is the recent changes happening around simple resource class selection.

**Comment by suskin** [comment]
> Which cloud provider were you using?

I was using GCP, and I was also using Crossplane 0.3 instead of master

File: bootstrap.sh (hunk 4 above)
Comment (suskin) [reply bootstrap.sh:13]: kk

File: pkg/trace/graphbuilder.go [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+
+	root, err = g.addNodeIfNotExist(u)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	err = g.fetchObj(root)
+	if kerrors.IsNotFound(err) {
+		return root, nil, errors.New(
+			fmt.Sprintf("Object to trace is not found: \"%s\" \"%s\" in namespace \"%s\"", groupRes, name, namespace))
```
Comment (suskin) [inline pkg/trace/graphbuilder.go]: Grammar nitpick: `Object to trace not found` or `Object to trace was not found` would be better. I would normally go with the `Object to trace not found` form

File: pkg/trace/graphbuilder_test.go [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+			args: args{
+				client:            fake.NewSimpleDynamicClient(scheme),
+				restMapper:        rm,
+				instancesToCreate: nil,
+				name:              "test",
+				namespace:         "testnamespace",
+				resource:          "kubernetescluster",
+			},
+			want: want{
+				err: errors.NewNotFound(schema.ParseGroupResource("kubernetesclusters.compute.crossplane.io"), "test"),
```
Comment (suskin) [inline pkg/trace/graphbuilder_test.go]: When I ran the unit tests, I got this:

```
--- FAIL: TestKubeGraphBuilder_BuildGraph (0.00s)
    --- FAIL: TestKubeGraphBuilder_BuildGraph/NotExist (0.00s)
        graphbuilder_test.go:306: g.BuildGraph(...): -want error, +got error:   interface{}(
            - 	e`kubernetesclusters.compute.crossplane.io "test" not found`,
            + 	e`Object to trace is not found: "kubernetescluster" "test" in namespace "testnamespace"`,
              )
```

Comment (turkenh) [reply pkg/trace/graphbuilder_test.go]: Fixed. I should get used to running make test before pushing :)

Comment (suskin) [reply pkg/trace/graphbuilder_test.go]: haha : ) - ideally we'll also set up some automated builds so we don't need to always remember

**Review by suskin — APPROVED** [review]
Thanks @turkenh ! This looks good to me. I believe we'll need to do a couple more things to get a release up, but we can work on it after this pull request is merged.
